src/LossFunctions.jl

```diff
@@ -14,22 +13,16 @@ import ..DimensionalAnalysisModule: violates_dimensional_constraints
 function _loss(
     x::AbstractArray{T}, y::AbstractArray{T}, loss::LT
 ) where {T<:DATA_TYPE,LT<:Union{Function,SupervisedLoss}}
-    if LT <: SupervisedLoss
-        return LossFunctions.mean(loss, x, y)
-    else
-        l(i) = loss(x[i], y[i])
-        return LossFunctions.mean(l, eachindex(x))
-    end
+    return sum(@. loss(x, y)) / length(x)
 end
 
 function _weighted_loss(
     x::AbstractArray{T}, y::AbstractArray{T}, w::AbstractArray{T}, loss::LT
 ) where {T<:DATA_TYPE,LT<:Union{Function,SupervisedLoss}}
     if LT <: SupervisedLoss
-        return LossFunctions.sum(loss, x, y, w; normalize=true)
+        return sum(@. loss(x, y) * w) / sum(w)
     else
-        l(i) = loss(x[i], y[i], w[i])
-        return sum(l, eachindex(x)) / sum(w)
+        return sum(@. loss(x, y, w)) / sum(w)
```

Comment:
Would be nice to avoid this, but it seems like `sum` is using the dataset array for temporary storage somehow? (Or Enzyme.jl thinks it is?)